src/main/java/com/google/api/generator/gapic/protowriter/Writer.java

```diff
@@ -43,52 +45,104 @@ public static CodeGeneratorResponse write(
       String outputFilePath) {
     ByteString.Output output = ByteString.newOutput();
     JavaWriterVisitor codeWriter = new JavaWriterVisitor();
-    JarOutputStream jos = null;
+    JarOutputStream jos;
     try {
       jos = new JarOutputStream(output);
     } catch (IOException e) {
       throw new GapicWriterException(e.getMessage());
     }
 
     for (GapicClass gapicClazz : clazzes) {
-      ClassDefinition clazz = gapicClazz.classDefinition();
+      String classPath = writeClazz(gapicClazz, codeWriter, jos);
+      writeSamples(gapicClazz, getSamplePackage(gapicClazz), classPath, jos);
+    }
+
+    writeMetadataFile(context, writePackageInfo(gapicPackageInfo, codeWriter, jos), jos);
+
+    try {
+      jos.finish();
+      jos.flush();
+    } catch (IOException e) {
+      throw new GapicWriterException(e.getMessage());
+    }
+
+    CodeGeneratorResponse.Builder response = CodeGeneratorResponse.newBuilder();
+    response
+        .setSupportedFeatures(CodeGeneratorResponse.Feature.FEATURE_PROTO3_OPTIONAL_VALUE)
+        .addFileBuilder()
+        .setName(outputFilePath)
+        .setContentBytes(output.toByteString());
+    return response.build();
+  }
 
-      clazz.accept(codeWriter);
-      String code = codeWriter.write();
-      codeWriter.clear();
+  private static String writeClazz(
+      GapicClass gapicClazz, JavaWriterVisitor codeWriter, JarOutputStream jos) {
+    ClassDefinition clazz = gapicClazz.classDefinition();
+
+    clazz.accept(codeWriter);
+    String code = codeWriter.write();
+    codeWriter.clear();
+
+    String path = getPath(clazz.packageString(), clazz.classIdentifier().name());
+    String className = clazz.classIdentifier().name();
+    JarEntry jarEntry = new JarEntry(String.format("%s/%s.java", path, className));
+    try {
+      jos.putNextEntry(jarEntry);
+      jos.write(code.getBytes(StandardCharsets.UTF_8));
+    } catch (IOException e) {
+      throw new GapicWriterException(
```

Comment:
Can we include the original exception as the cause of `GapicWriterException`? This is usually how we handle and re-throw exceptions because you can keep the stack trace. You would need an overloaded constructor for `GapicWriterException` but that's totally worth it.

Reply:
done